File: TLS Inspector/View Controllers/More/AboutViewController.m
#import "AboutViewController.h"
#import "AppLinks.h"
#import "TitleValueTableViewCell.h"

@interface AboutViewController () <UITableViewDataSource, UITableViewDelegate>

@property (weak, nonatomic) IBOutlet UITableView *tableView;
@property (strong, nonatomic) AppLinks * appLinks;

@end

@implementation AboutViewController

static NSString * PROJECT_GITHUB_URL = @"https://github.com/certificate-helper/TLS-Inspector/";
static NSString * PROJECT_URL = @"https://tlsinspector.com/";
static NSString * PROJECT_CONTRIBUTE_URL = @"https://tlsinspector.com/contribute.html";
static NSString * PROJECT_TESTFLIGHT_APPLICATION = @"https://tlsinspector.com/beta.html";

- (void) viewDidLoad {
    [super viewDidLoad];
    [self setNeedsStatusBarAppearanceUpdate];
    self.appLinks = [AppLinks new];
}

- (UIStatusBarStyle) preferredStatusBarStyle {
    return UIStatusBarStyleLightContent;
}

- (void) didReceiveMemoryWarning {
    [super didReceiveMemoryWarning];
}

- (IBAction) closeButton:(UIButton *)sender {
    [self dismissViewControllerAnimated:YES completion:nil];
}

# pragma mark - Table View Source

- (NSInteger) numberOfSectionsInTableView:(UITableView *)tableView {
    return 2;
}

- (NSInteger) tableView:(UITableView *)tableView numberOfRowsInSection:(NSInteger)section {
    switch (section) {
        case 0:
            return 3;
        case 1:
            return 1;
    }
    return 0;
}

- (UITableViewCell *) tableView:(UITableView *)tableView cellForRowAtIndexPath:(NSIndexPath *)indexPath {
    if (indexPath.section == 0) {
        UITableViewCell * cell = [tableView dequeueReusableCellWithIdentifier:@"action" forIndexPath:indexPath];
        UILabel * label = (UILabel *)[cell viewWithTag:1];
        label.textColor = themeTextColor;
        switch (indexPath.row) {
            case 0:
                label.text = l(@"Tell Friends About TLS Inspector");
                break;
            case 1:
                label.text = l(@"Rate TLS Inspector in the App Store");
                break;
            case 2:
                label.text = l(@"Submit Feedback");
                break;
        }
        return cell;
    } else if (indexPath.section == 1) {
        UITableViewCell * cell = [tableView dequeueReusableCellWithIdentifier:@"action" forIndexPath:indexPath];
        UILabel * label = (UILabel *)[cell viewWithTag:1];
        label.textColor = themeTextColor;
        switch (indexPath.row) {
            case 0:
                label.text = l(@"Contribute to TLS Inspector");
                break;
            case 1:
                label.text = l(@"Test New Features");
                break;
        }
        return cell;
    }
    return nil;
}

- (void) tableView:(UITableView *)tableView didSelectRowAtIndexPath:(NSIndexPath *)indexPath {
    if (indexPath.section == 0 && indexPath.row == 0) {
        NSString * blurb = format(@"Trust & Safety On-the-go with TLS Inspector: %@", PROJECT_URL);
        UIActivityViewController *activityController = [[UIActivityViewController alloc]
                                                        initWithActivityItems:@[blurb]
                                                        applicationActivities:nil];
        activityController.popoverPresentationController.sourceView = [tableView cellForRowAtIndexPath:indexPath];
        [self presentViewController:activityController animated:YES completion:nil];
    } else if (indexPath.section == 0 && indexPath.row == 1) {
        [self.appLinks showAppInAppStorInViewController:self dismissed:nil];
    } else if (indexPath.section == 0 && indexPath.row == 2) {
        [uihelper
         presentActionSheetInViewController:self
         attachToTarget:[ActionTipTarget targetWithView:[tableView cellForRowAtIndexPath:indexPath]]
         title:l(@"What kind of feedback would you like to submit?")
         subtitle:l(@"All feedback is appreciated!")
         cancelButtonTitle:l(@"Cancel")
         items:@[
                 l(@"Report Issue on GitHub"),
                 l(@"Contact Us")
                 ]
         dismissed:^(NSInteger selectedIndex) {
             switch (selectedIndex) {
                 case 0:
                     open_url(nstrcat(PROJECT_GITHUB_URL, @"issues/new"));
                     break;
                 case 1: {
                     [self.appLinks showEmailComposeSheetForAppInViewController:self dismissed:nil];
                     break;
                 } default:
                     break;
             }
         }];
    } else if (indexPath.section == 1 && indexPath.row == 0) {
        open_url(PROJECT_CONTRIBUTE_URL);
    } else if (indexPath.section == 1 && indexPath.row == 1) {
        open_url(PROJECT_TESTFLIGHT_APPLICATION);
    }
}

- (NSString *) tableView:(UITableView *)tableView titleForHeaderInSection:(NSInteger)section {
    switch (section) {
        case 0:
            return l(@"Share & Feedback");
        case 1:
            return l(@"Get Involved");
    }
    return @"";
}

- (NSString *) tableView:(UITableView *)tableView titleForFooterInSection:(NSInteger)section {
    switch (section) {
        case 0: {
            NSDictionary * infoDictionary = [[NSBundle mainBundle] infoDictionary];
            return format(@"App: %@ (%@). OpenSSL: %@, cURL: %@",
                          [infoDictionary objectForKey:@"CFBundleShortVersionString"],
                          [infoDictionary objectForKey:(NSString *)kCFBundleVersionKey],
                          [CertificateKit opensslVersion],
                          [CertificateKit libcurlVersion]);
        } case 1:
            return l(@"TLS Inspector is Free and Libre software licensed under GNU GPLv3. TLS Inspector is copyright © 2016-2018 Ian Spence.");
    }
    return @"";
}

@end

File: tls-inspector/View Controllers/AboutViewController.swift
import UIKit
import CertificateKit
import WebKit

class AboutTableViewController: UIViewController, UITableViewDataSource, UITableViewDelegate {
    let projectGithubURL = "https://tlsinspector.com/github.html"
    let projectURL = "https://tlsinspector.com/"
    let projectContributeURL = "https://tlsinspector.com/contribute.html"
    let testflightURL = "https://tlsinspector.com/beta.html"
    let mastodonURL = "https://infosec.exchange/@tlsinspector"
    @IBOutlet weak var lockCircle: UIImageView!
    @IBOutlet weak var tableView: UITableView!
    var quotes: [String] = []
    var taps = 0
    var sections: [TableViewSection] = []

    override func viewDidLoad() {
        super.viewDidLoad()

        if let q = loadQuotes() {
            self.quotes = q
        }

        let tap = UITapGestureRecognizer(target: self, action: #selector(self.didTapPicture(target:)))
        self.lockCircle.isUserInteractionEnabled = true
        self.lockCircle.addGestureRecognizer(tap)

        self.navigationItem.leftBarButtonItem = UIBarButtonItem(barButtonSystemItem: .stop, target: self, action: #selector(dismissView))

        self.sections = [
            buildShareSection(),
            buildGetInvolvedSection()
        ]

        if #available(iOS 15, *) {
            self.sections.append(buildMoreFromSection())
        }
    }

    @objc func dismissView(_ sendor: Any?) {
        self.dismiss(animated: true)
    }

    func buildShareSection() -> TableViewSection {
        let section = TableViewSection()
        section.title = lang(key: "Share & Feedback")
        let opensslVersion = CertificateKit.opensslVersion()
        let libcurlVersion = CertificateKit.libcurlVersion()
        section.footer = "App: \(AppInfo.version()) (\(AppInfo.build())), OpenSSL: \(opensslVersion), curl: \(libcurlVersion)"

        let shareCell = TableViewCell(UITableViewCell())
        shareCell.cell.textLabel?.text = lang(key: "Share TLS Inspector")
        shareCell.didSelect = { _, _ in
            let blurb = lang(key: "Trust & Safety On-The-Go with TLS Inspector: {url}", args: [self.projectURL])
            let activityController = UIActivityViewController(activityItems: [blurb], applicationActivities: nil)
            ActionTipTarget(view: shareCell.cell).attach(to: activityController.popoverPresentationController)
            self.present(activityController, animated: true, completion: nil)
        }

        let rateCell = TableViewCell(UITableViewCell())
        rateCell.cell.textLabel?.text = lang(key: "Rate in App Store")
        rateCell.didSelect = { _, _ in
            AppLinks.current.showAppStoreIn(self, appId: AppLinks.tlsInspectorAppId, dismissed: nil)
        }

        var cells = [
            shareCell,
            rateCell
        ]

        if #available(iOS 15, *) {
            let feedbackCell = TableViewCell(UITableViewCell())
            feedbackCell.cell.textLabel?.text = lang(key: "Provide Feedback")
            feedbackCell.didSelect = { _, _ in
                ContactTableViewController.show(self) { (support) in
                    AppLinks.current.showEmailCompose(viewController: self, object: support, includeLogs: false, dismissed: nil)
                }
            }
            cells.append(feedbackCell)
        }

        section.cells = cells
        return section
    }

    func buildGetInvolvedSection() -> TableViewSection {
        let section = TableViewSection()
        section.title = lang(key: "Get Involved")
        section.footer = lang(key: "copyright_license_footer")

        let followCell = TableViewCell(UITableViewCell())
        followCell.cell.textLabel?.text = lang(key: "Follow @tlsinspector on Mastodon")
        followCell.cell.imageView?.image = UIImage(named: "Mastodon")
        followCell.didSelect = { _, _ in
            OpenURLInSafari(self.mastodonURL)
        }

        let contributeCell = TableViewCell(UITableViewCell())
        contributeCell.cell.textLabel?.text = lang(key: "Contribute to TLS Inspector")
        contributeCell.didSelect = { _, _ in
            OpenURLInSafari(self.projectContributeURL)
        }

        let attributeCell = TableViewCell(UITableViewCell())
        attributeCell.cell.textLabel?.text = lang(key: "Open Source Licenses & Attributions")
        attributeCell.didSelect = { _, _ in
            guard let attrPath = Bundle.main.url(forResource: "attr", withExtension: "html") else { return }
            guard let attrHtml = try? String(contentsOf: attrPath) else { return }

            let viewController = UIViewController()
            let webView = WKWebView()
            viewController.view = webView
            viewController.title = "Open Source Licenses & Attributions"
            webView.loadHTMLString(attrHtml, baseURL: nil)
            self.navigationController?.pushViewController(viewController, animated: true)
        }

        section.cells = [
            followCell,
            contributeCell,
            attributeCell,
        ]

        return section
    }

    func buildMoreFromSection() -> TableViewSection {
        let section = TableViewSection()

        section.title = lang(key: "More from the developer")

        let dnsInspectorCell = TableViewCell(UITableViewCell())
        dnsInspectorCell.cell.imageView?.image = UIImage(named: "DNS Inspector Icon")
        dnsInspectorCell.cell.textLabel?.text = "DNS Inspector"
        dnsInspectorCell.didSelect = { _, _ in
            AppLinks.current.showAppStoreIn(self, appId: AppLinks.dnsInspectorAppId, dismissed: nil)
        }

        section.cells.append(dnsInspectorCell)

        return section
    }

    func loadQuotes() -> [String]? {
        guard let quotesPath = Bundle.main.path(forResource: "QuoteList", ofType: "plist") else { return nil }
        guard let quotes = NSArray.init(contentsOfFile: quotesPath) as? [String] else { return nil }
        return quotes
    }

    @objc func didTapPicture(target: UIImageView) {
        if self.quotes.count == 0 {
            return
        }

        self.taps += 1
        if self.taps >= 3 {
            self.taps = 0

            UIHelper(self).presentAlert(title: self.quotes.randomElement() ?? "", body: "", dismissed: nil)
        }
    }

    // MARK: - Table view data source
    func numberOfSections(in tableView: UITableView) -> Int {
        return self.sections.count
    }

    func tableView(_ tableView: UITableView, numberOfRowsInSection section: Int) -> Int {
        return self.sections[section].cells.count
    }

    func tableView(_ tableView: UITableView, cellForRowAt indexPath: IndexPath) -> UITableViewCell {
        return self.sections[indexPath.section].cells[indexPath.row].cell
    }

    func tableView(_ tableView: UITableView, titleForHeaderInSection section: Int) -> String? {
        return self.sections[section].title
    }

    func tableView(_ tableView: UITableView, titleForFooterInSection section: Int) -> String? {
        return self.sections[section].footer
    }

    func tableView(_ tableView: UITableView, didSelectRowAt indexPath: IndexPath) {
        guard let action = self.sections[indexPath.section].cells[indexPath.row].didSelect else { return }
        action(tableView, indexPath)
    }
}
